### DISCUSSION TITLE
Modelling an onlap surface across more than one series

### DISCUSSION THREAD

**Discussion by lucianomarquetto** [gempy-project/gempy#784] [General]
Hi all,

I am currently working on a coastal area with 5 series of horizontal stratigraphy, but with the basement outcropping in the northwest. I am trying to model an onlap surface to represent this basement coming from depth and outcropping, but can't get it to work (the horizontal layers do not compute properly). 

This is my stratigraphy:

![image](https://user-images.githubusercontent.com/103533248/229872241-a68fdea0-8238-473b-815c-960ed558fb32.png)

I was able to model the basement satisfactorily:

![image](https://user-images.githubusercontent.com/103533248/229866070-a4678eb0-b3a9-4916-a13f-2f1e6328bb4f.png)

And the relation to the horizontal stratigraphy I'm trying to reach is something like below:

![image](https://user-images.githubusercontent.com/103533248/229867663-38568cf6-fd48-455e-930c-40c81bd2993c.png)

What I have already tried:

- Setting all bottom relationships as 'onlap' (as in https://github.com/cgre-aachen/gempy/issues/564);
- Changing the stratigraphical order of the onlap surface up and down;
- Adding onlap surfaces intercalated with my original series (using the same points for all onlap surfaces and setting all bottom surfaces to 'onlap');
- Keeping all my horizontal stratigraphy together in the 'Default Series' instead of one series each;
- Adding a second onlap surface in the middle or top of the horizontal series (using the same points as the original).

I am able to get variable but not satisfactory results. Any suggestions? Is there a code function that allows an onlap surface to cross all series above it instead of only the series right above?

I am using gempy 2.2.11 in Ubuntu 22.04.

Thanks!

**Comment by javoha** [comment]
Hi, 
would you be able to provide your input dataset - that makes it a lot easier to figure out the issue.
Cheers
Jan

**Comment by lucianomarquetto** [comment]
Hi Jan! Thanks for responding - please see the data I'm using below!

[Troian_2020_gempy_input_data.csv](https://github.com/cgre-aachen/gempy/files/11174166/Troian_2020_gempy_input_data.csv)

Cheers,

Luciano

PS: Edited the csv file to add a few points and consider for terrain elevation in the subsurface points.

**Reply by javoha** [reply]
Is that a mix of surface points and orientations? Looks like all orientations on first glance.

**Reply by lucianomarquetto** [reply]
Hi Jan,

I have the surfaces and orientations on the same file (same number of orientations and surfaces). I do not know if this is not a good practice, but I considered all my horizontal stratigraphy points (units 1, 2, 3, 4 and pelotas_lower) to have orientations (so azimuth = 0, dip = 0 and polarity = 1 for each), and my onlap surface between those and the basement was a trial and error to get to the point seen in the image I sent earlier in the discussion. The onlap surfaces named "dummy_point" are the outcropping basement, which I defined based on topography in QGIS.

Hope that helps, otherwise please let me know and I will share more info!

Cheers,

Luciano

**Comment by javoha** [comment]
Hey @lucianomarquetto,
maybe I solved this, I think your bottom relations where a little out of order - it can be very confusing in gempy. How does this look to you (vertical exaggeration is 50, sorry for the low resolution):
![Model](https://user-images.githubusercontent.com/38279459/230641003-8c15f1e9-b8fb-4585-b15c-60e46f20840b.PNG)

I removed the "pelotas lower" points, as those looked like they were in the dataframe twice. You will achieve the correct outcropping (geological map) by adding a topography. I renamed the series a little for my convenience and set the bottom relation of the horizontal strata series to "Onlap". 

```Python
gp.map_stack_to_surfaces(geo_model, {"Horizontal_Series": ('unit4', 'unit3','unit2', 'unit1'),
                                     "Onlap_Series": ('onlap_surface5'),
                                     "Basement_Series":('basement')})

geo_model.set_bottom_relation("Horizontal_Series", "Onlap")
```

Is this what you were looking for?

**Comment by lucianomarquetto** [comment]
Hi @javoha! This is pretty much what I am looking for at the present, thanks! Based on your comment I was able to reproduce the model in a similar way (took the pelotas_lower out and it worked).

I think my hassle came from a confusion I made about onlap surface vs regular surface - I added the "pelotas_lower" to represent the undifferentiated strata below my four units, and the "onlap surface" to define the boundary between the horizontal strata (pelotas_lower and units 1-4) and the basement (so the "pelotas_lower" had fewer points than the "onlap surface"). I understand by reading your comment that having only the onlap surface is enough for this.

I went further, based on your code, and defined my four units as one series each, which also worked now. In the future I plan to detail them, so better to already try how it would work.

Anyway, one thing that caught my attention now is that my 3D plots are odd. At first I thought I was running something different from you, because in your model I could see the basement (red) in the northwest and the onlap surface (purple) in the southeast, and all surfaces seemed to be in the right depth comparing to the translucent blocks. In my plots (both defining the four units as one series only or one series each), my southeast section becomes red as if basement, and surfaces seem off compared to the translucent block limits. But then I noticed my top surface, which should be blue, is appearing with the color of the second surface:

![image](https://user-images.githubusercontent.com/103533248/231264711-f3e9d1ca-1f15-4e4c-95db-a9166b64972b.png)

I believe this is then a pyvista problem? My pyvista is 0.37.0, and I tried plotting as basic and background running `gp.plot_3d(geo_model, ve=100, plotter_type='background')` with no difference. Tried also updating it to 0.38.5 in one of my virtual environments, but now I got some compatibility problem I'm trying to solve.

Anyway, I will try finding a solution to this color and surface issue, and if you happen to know what I can do please share. So far I can only thank you and the Gempy community for all help and support so far, it has been an enriching experience to work with the package!

**Comment by lucianomarquetto** [comment]
Ok just noticed this: https://github.com/cgre-aachen/gempy/issues/754

Probably my pyvista problems are related to my version then!

**Reply by javoha** [reply]
Yeah I get the same problem with the colors - for the image I accidentally used one of my older environments where everything was still working properly. Hopefully most of these dependency related issues will be fixed in the next rellease, but of course this will always be part of open-source development. If its super urgent for you to fix this, you could try to export the gempy surfaces and work directly with pyvista to create "your own" plots, but this will probably take some time. 
Thanks for the nice words, always appreciated!

Edit: I will actually reopen this, so people can find it easier when looking for help, maybe you can mark it as answered.

**Reply by lucianomarquetto** [reply]
Hey Jan,

I am not in a hurry to have it right, but will definitely try exporting the model and working directly on pyvista if needed, thanks!

Further investigating I noticed it is not pyvista (at least not alone) that is giving me those off surfaces and color problems. I tried creating a new virtual environment with python 3.8.1 and getting the gempy required packages from a requirements.txt file running `pip install -r requirements.txt`. I used the 'requires.txt' from the tar.gz file in https://pypi.org/project/gempy/#files, just changed the pyvista version from the original file (pyvista>=0.25) to pyvista==0.34.2 based on https://github.com/cgre-aachen/gempy/issues/754. Nonetheless, I am still having the same visualization issues. 

Would you mind sharing your 'older environment' yml file so I can try reproducing it here? Now that I know it is just the 3D visualization and not the model that is odd I can keep working with no problems, but still would like to try and have a better visualization if I can, until the next gempy release.

Cheers!

**Reply by javoha** [reply]
Hi, 
here is the environment I used, it's really quite old though - not sure if it helps. For some reason I cant upload a .yml file here in the comment, so here what i get from conda env export:

name: gempy_env
channels:
  - defaults
dependencies:
  - _anaconda_depends=2020.02=py37_0
  - alabaster=0.7.12=py37_0
  - anaconda=custom=py37_1
  - anaconda-client=1.7.2=py37_0
  - anaconda-project=0.8.4=py_0
  - argh=0.26.2=py37_0
  - asn1crypto=1.3.0=py37_0
  - astroid=2.3.3=py37_0
  - astropy=4.0.1.post1=py37he774522_1
  - atomicwrites=1.4.0=py_0
  - attrs=19.3.0=py_0
  - autopep8=1.4.3=py37_0
  - babel=2.8.0=py_0
  - backcall=0.2.0=py_0
  - backports=1.0=py_2
  - backports.shutil_get_terminal_size=1.0.0=py37_2
  - bcrypt=3.1.7=py37he774522_1
  - beautifulsoup4=4.9.1=py37_0
  - bitarray=1.2.2=py37he774522_0
  - bkcharts=0.2=py37_0
  - blas=1.0=mkl
  - bleach=3.1.5=py_0
  - blosc=1.19.0=h7bd577a_0
  - bokeh=2.1.0=py37_0
  - boto=2.49.0=py37_0
  - bottleneck=1.3.2=py37h2a96729_1
  - brotlipy=0.7.0=py37he774522_1000
  - bzip2=1.0.8=he774522_0
  - ca-certificates=2021.10.26=haa95532_2
  - certifi=2021.10.8=py37haa95532_0
  - cffi=1.14.0=py37h7a1dbc1_0
  - chardet=3.0.4=py37_1003
  - click=7.1.2=py_0
  - cloudpickle=1.4.1=py_0
  - clyent=1.2.2=py37_1
  - colorama=0.4.3=py_0
  - comtypes=1.1.7=py37_1001
  - console_shortcut=0.1.1=4
  - contextlib2=0.6.0.post1=py_0
  - cryptography=2.9.2=py37h7a1dbc1_0
  - curl=7.71.1=h86230a5_2
  - cycler=0.10.0=py37_0
  - cython=0.29.20=py37ha925a31_0
  - cytoolz=0.10.1=py37he774522_0
  - dask=2.19.0=py_0
  - dask-core=2.19.0=py_0
  - decorator=4.4.2=py_0
  - defusedxml=0.6.0=py_0
  - diff-match-patch=20181111=py_0
  - distributed=2.19.0=py37_0
  - docutils=0.16=py37_1
  - entrypoints=0.3=py37_0
  - et_xmlfile=1.0.1=py37_0
  - expat=2.4.1=h6c2663c_2
  - fastcache=1.1.0=py37he774522_0
  - filelock=3.0.12=py_0
  - flake8=3.7.9=py37_0
  - flask=1.1.2=py_0
  - freetype=2.10.2=hd328e21_0
  - freexl=1.0.6=h2bbff1b_0
  - fsspec=0.7.4=py_0
  - future=0.18.2=py37_1
  - gdal=2.3.3=py37hdf43c64_0
  - geos=3.7.1=h33f27b4_0
  - get_terminal_size=1.0.0=h38e98db_0
  - gevent=1.4.0=py37he774522_0
  - glob2=0.7=py_0
  - gmpy2=2.0.8=py37h0964b28_3
  - greenlet=0.4.16=py37he774522_0
  - h5py=2.10.0=py37h5e291fa_0
  - hdf4=4.2.13=h712560f_2
  - hdf5=1.10.4=h7ebc959_0
  - heapdict=1.0.1=py_0
  - html5lib=1.0.1=py37_0
  - hypothesis=5.16.1=py_0
  - icc_rt=2019.0.0=h0cc432a_1
  - icu=58.2=ha925a31_3
  - idna=2.9=py_1
  - imageio=2.8.0=py_0
  - imagesize=1.2.0=py_0
  - importlib-metadata=1.6.1=py37_0
  - importlib_metadata=1.6.1=0
  - intel-openmp=2020.1=216
  - intervaltree=3.0.2=py_0
  - ipykernel=5.3.0=py37h5ca1d4c_0
  - ipython=7.15.0=py37_0
  - ipython_genutils=0.2.0=py37_0
  - ipywidgets=7.5.1=py_0
  - isort=4.3.21=py37_0
  - itsdangerous=1.1.0=py37_0
  - jdcal=1.4.1=py_0
  - jedi=0.15.2=py37_0
  - jinja2=2.11.2=py_0
  - joblib=0.15.1=py_0
  - jpeg=9b=hb83a4c4_2
  - json5=0.9.5=py_0
  - jsonschema=3.2.0=py37_0
  - jupyter=1.0.0=py37_7
  - jupyter_client=6.1.3=py_0
  - jupyter_console=6.1.0=py_0
  - jupyter_core=4.6.3=py37_0
  - jupyterlab=2.1.4=py_1
  - jupyterlab_server=1.1.5=py_0
  - kealib=1.4.7=h07cbb95_6
  - keyring=21.1.1=py37_2
  - kiwisolver=1.2.0=py37h74a9793_0
  - krb5=1.16.4=hc04afaa_0
  - lazy-object-proxy=1.5.0=py37he774522_0
  - libarchive=3.4.2=h5e25573_0
  - libboost=1.67.0=hd9e427e_4
  - libcurl=7.71.1=h86230a5_2
  - libgdal=2.3.3=h10f50ba_0
  - libgpuarray=0.7.6=hfa6e2cd_0
  - libiconv=1.15=h1df5818_7
  - libkml=1.3.0=he5f2a48_4
  - liblief=0.10.1=ha925a31_0
  - libnetcdf=4.6.1=h411e497_2
  - libpng=1.6.37=h2a8f88b_0
  - libpq=11.2=h3235a2c_0
  - libpython=2.1=py37_0
  - libsodium=1.0.16=h9d3ae62_0
  - libspatialindex=1.9.3=h33f27b4_0
  - libspatialite=4.3.0a=hc36aec2_19
  - libssh2=1.9.0=h7a1dbc1_1
  - libtiff=4.1.0=h56a325e_1
  - libxml2=2.9.10=h464c3ec_1
  - libxslt=1.1.34=he774522_0
  - llvmlite=0.32.1=py37ha925a31_0
  - locket=0.2.0=py37_1
  - lxml=4.5.1=py37h1350720_0
  - lz4-c=1.9.2=h62dcd97_0
  - lzo=2.10=he774522_2
  - m2w64-binutils=2.25.1=5
  - m2w64-bzip2=1.0.6=6
  - m2w64-crt-git=5.0.0.4636.2595836=2
  - m2w64-gcc=5.3.0=6
  - m2w64-gcc-ada=5.3.0=6
  - m2w64-gcc-fortran=5.3.0=6
  - m2w64-gcc-libgfortran=5.3.0=6
  - m2w64-gcc-libs=5.3.0=7
  - m2w64-gcc-libs-core=5.3.0=7
  - m2w64-gcc-objc=5.3.0=6
  - m2w64-gmp=6.1.0=2
  - m2w64-headers-git=5.0.0.4636.c0ad18a=2
  - m2w64-isl=0.16.1=2
  - m2w64-libiconv=1.14=6
  - m2w64-libmangle-git=5.0.0.4509.2e5a9a2=2
  - m2w64-libwinpthread-git=5.0.0.4634.697f757=2
  - m2w64-make=4.1.2351.a80a8b8=2
  - m2w64-mpc=1.0.3=3
  - m2w64-mpfr=3.1.4=4
  - m2w64-pkg-config=0.29.1=2
  - m2w64-toolchain=5.3.0=7
  - m2w64-tools-git=5.0.0.4592.90b8472=2
  - m2w64-windows-default-manifest=6.4=3
  - m2w64-winpthreads-git=5.0.0.4634.697f757=2
  - m2w64-zlib=1.2.8=10
  - mako=1.1.3=py_0
  - markupsafe=1.1.1=py37he774522_0
  - matplotlib=3.2.2=0
  - matplotlib-base=3.2.2=py37h64f37c6_0
  - mccabe=0.6.1=py37_1
  - menuinst=1.4.16=py37he774522_1
  - mistune=0.8.4=py37he774522_0
  - mkl=2020.1=216
  - mkl-service=2.3.0=py37hb782905_0
  - mkl_fft=1.1.0=py37h45dec08_0
  - mkl_random=1.1.1=py37h47e9c7a_0
  - mock=4.0.2=py_0
  - more-itertools=8.4.0=py_0
  - mpc=1.1.0=h7edee0f_1
  - mpfr=4.0.2=h62dcd97_1
  - mpir=3.0.0=hec2e145_1
  - mpmath=1.1.0=py37_0
  - msgpack-python=1.0.0=py37h74a9793_1
  - msys2-conda-epoch=20160418=1
  - multipledispatch=0.6.0=py37_0
  - nbconvert=5.6.1=py37_0
  - nbformat=5.0.7=py_0
  - networkx=2.4=py_0
  - nltk=3.5=py_0
  - nose=1.3.7=py37_2
  - notebook=6.0.3=py37_0
  - numba=0.49.1=py37h47e9c7a_0
  - numexpr=2.7.1=py37h25d0782_0
  - numpy-base=1.18.5=py37hc3f5095_0
  - numpydoc=1.0.0=py_0
  - olefile=0.46=py37_0
  - openpyxl=3.0.3=py_0
  - openssl=1.1.1l=h2bbff1b_0
  - packaging=20.4=py_0
  - pandoc=2.2.3.2=0
  - pandocfilters=1.4.2=py37_1
  - paramiko=2.7.1=py_0
  - parso=0.5.2=py_0
  - partd=1.1.0=py_0
  - path=13.1.0=py37_0
  - path.py=12.4.0=0
  - pathlib2=2.3.5=py37_0
  - pathtools=0.1.2=py_1
  - patsy=0.5.1=py37_0
  - pcre=8.45=hd77b12b_0
  - pep8=1.7.1=py37_0
  - pexpect=4.8.0=py37_0
  - pickleshare=0.7.5=py37_0
  - pillow=7.1.2=py37hcc1f983_0
  - pip=20.1.1=py37_1
  - pkginfo=1.5.0.1=py37_0
  - pluggy=0.13.1=py37_0
  - ply=3.11=py37_0
  - powershell_shortcut=0.0.1=3
  - proj4=5.2.0=ha925a31_1
  - prometheus_client=0.8.0=py_0
  - prompt-toolkit=3.0.5=py_0
  - prompt_toolkit=3.0.5=0
  - psutil=5.7.0=py37he774522_0
  - py=1.8.2=py_0
  - py-lief=0.10.1=py37ha925a31_0
  - pycodestyle=2.5.0=py37_0
  - pycosat=0.6.3=py37he774522_0
  - pycparser=2.20=py_0
  - pycrypto=2.6.1=py37he774522_10
  - pycurl=7.43.0.5=py37h7a1dbc1_0
  - pydocstyle=5.0.2=py_0
  - pyflakes=2.1.1=py37_0
  - pygments=2.6.1=py_0
  - pygpu=0.7.6=py37h2a96729_0
  - pylint=2.4.4=py37_0
  - pynacl=1.3.0=py37h62dcd97_0
  - pyodbc=4.0.30=py37ha925a31_0
  - pyopenssl=19.1.0=py37_0
  - pyparsing=2.4.7=py_0
  - pyqt=5.9.2=py37h6538335_2
  - pyreadline=2.1=py37_1
  - pyrsistent=0.16.0=py37he774522_0
  - pysocks=1.7.1=py37_0
  - pytables=3.6.1=py37h1da0976_0
  - pytest=5.4.3=py37_0
  - pytest-arraydiff=0.3=py37h39e3cac_0
  - pytest-astropy=0.8.0=py_0
  - pytest-astropy-header=0.1.2=py_0
  - pytest-doctestplus=0.7.0=py_0
  - pytest-openfiles=0.5.0=py_0
  - pytest-remotedata=0.3.2=py37_0
  - python=3.7.7=h81c818b_4
  - python-dateutil=2.8.1=py_0
  - python-jsonrpc-server=0.3.4=py_0
  - python-language-server=0.31.10=py37_0
  - python-libarchive-c=2.9=py_0
  - pytz=2020.1=py_0
  - pywavelets=1.1.1=py37he774522_0
  - pywin32=227=py37he774522_1
  - pywin32-ctypes=0.2.0=py37_1000
  - pywinpty=0.5.7=py37_0
  - pyyaml=5.3.1=py37he774522_0
  - pyzmq=18.1.1=py37ha925a31_0
  - qdarkstyle=2.8.1=py_0
  - qt=5.9.7=vc14h73c81de_0
  - qtawesome=0.7.2=py_0
  - qtconsole=4.7.5=py_0
  - qtpy=1.9.0=py_0
  - regex=2020.5.14=py37he774522_0
  - requests=2.24.0=py_0
  - rope=0.17.0=py_0
  - rtree=0.9.4=py37h21ff451_1
  - ruamel_yaml=0.15.87=py37he774522_0
  - scikit-learn=0.23.1=py37h25d0782_0
  - scipy=1.4.1=py37h9439919_0
  - seaborn=0.10.1=py_0
  - send2trash=1.5.0=py37_0
  - setuptools=47.3.1=py37_0
  - simplegeneric=0.8.1=py37_2
  - singledispatch=3.4.0.3=py37_0
  - sip=4.19.8=py37h6538335_0
  - six=1.15.0=py_0
  - snappy=1.1.7=h777316e_3
  - snowballstemmer=2.0.0=py_0
  - sortedcollections=1.2.1=py_0
  - sortedcontainers=2.2.2=py_0
  - soupsieve=2.0.1=py_0
  - sphinx=3.1.1=py_0
  - sphinxcontrib=1.0=py37_1
  - sphinxcontrib-applehelp=1.0.2=py_0
  - sphinxcontrib-devhelp=1.0.2=py_0
  - sphinxcontrib-htmlhelp=1.0.3=py_0
  - sphinxcontrib-jsmath=1.0.1=py_0
  - sphinxcontrib-qthelp=1.0.3=py_0
  - sphinxcontrib-serializinghtml=1.1.4=py_0
  - sphinxcontrib-websupport=1.2.2=py_0
  - spyder=4.1.3=py37_0
  - spyder-kernels=1.9.1=py37_0
  - sqlalchemy=1.3.17=py37he774522_0
  - sqlite=3.32.3=h2a8f88b_0
  - statsmodels=0.11.1=py37he774522_0
  - sympy=1.6=py37_0
  - tbb=2020.0=h74a9793_0
  - tblib=1.6.0=py_0
  - terminado=0.8.3=py37_0
  - testpath=0.4.4=py_0
  - theano=1.0.4=py37_0
  - threadpoolctl=2.1.0=pyh5ca1d4c_0
  - tk=8.6.10=he774522_0
  - toolz=0.10.0=py_0
  - tornado=6.0.4=py37he774522_1
  - tqdm=4.46.1=py_0
  - traitlets=4.3.3=py37_0
  - typing_extensions=3.7.4.2=py_0
  - ujson=1.35=py37hfa6e2cd_0
  - unicodecsv=0.14.1=py37_0
  - urllib3=1.25.9=py_0
  - vc=14.1=h0510ff6_4
  - vs2015_runtime=14.16.27012=hf0eaf9b_2
  - watchdog=0.10.2=py37_0
  - wcwidth=0.2.4=py_0
  - webencodings=0.5.1=py37_1
  - werkzeug=1.0.1=py_0
  - wheel=0.34.2=py37_0
  - widgetsnbextension=3.5.1=py37_0
  - win_inet_pton=1.1.0=py37_0
  - win_unicode_console=0.5=py37_0
  - wincertstore=0.2=py37_0
  - winpty=0.4.3=4
  - wrapt=1.12.1=py37he774522_1
  - xerces-c=3.2.3=ha925a31_0
  - xlrd=1.2.0=py37_0
  - xlsxwriter=1.2.9=py_0
  - xlwings=0.19.4=py37_0
  - xlwt=1.3.0=py37_0
  - xz=5.2.5=h62dcd97_0
  - yaml=0.1.7=hc54c509_2
  - yapf=0.29.0=py_0
  - zeromq=4.3.1=h33f27b4_3
  - zict=2.0.0=py_0
  - zipp=3.1.0=py_0
  - zlib=1.2.11=h62dcd97_4
  - zstd=1.4.4=ha9fde0e_3
  - pip:
    - appdirs==1.4.4
    - autograd==1.3
    - meshio==4.0.15
    - numpy==1.18.5
    - pandas==1.3.4
    - pyvista==0.24.3
    - scikit-image==0.17.2
    - scooby==0.5.5
    - sphinx-gallery==0.7.0
    - tifffile==2020.6.3
    - vtk==9.0.0

**Reply by lucianomarquetto** [reply]
Hi Jan,

Thanks for sharing your environment - I will try to reproduce it or at least find the specific packages related to 3D visualization and compare to my versions. It sure will help me go further in this matter! When (if) I reach any conclusion will share here!

Cheers,

Luciano.
